File: AndroidWebBox/coconut-core/src/main/java/com/sniper/coconut/bridge/SecurityAuditLog.kt
package com.sniper.coconut.bridge

import com.sniper.coconut.utils.Logger
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Security Audit Log
 *
 * Singleton in coconut-core to record security rejection events.
 * SecurityComponent in coconut-plugins reads from this.
 */
object SecurityAuditLog {

    private const val TAG = "SecurityAuditLog"
    private const val MAX_ENTRIES = 500

    data class AuditEntry(
        val eventType: String,
        val method: String,
        val requestId: String,
        val detail: String,
        val timestamp: Long = System.currentTimeMillis()
    )

    private val entries = CopyOnWriteArrayList<AuditEntry>()

    // Event types
    const val EVENT_DOMAIN_REJECTED = "domain_rejected"
    const val EVENT_RATE_LIMITED = "rate_limited"
    const val EVENT_PARAMS_OVERSIZED = "params_oversized"
    const val EVENT_SIGNATURE_INVALID = "signature_invalid"
    const val EVENT_SIGNATURE_EXPIRED = "signature_expired"
    const val EVENT_NONCE_REUSED = "nonce_reused"
    const val EVENT_TOKEN_INVALID = "token_invalid"

    /**
     * Record a security event
     */
    fun record(eventType: String, method: String, requestId: String, detail: String) {
        if (entries.size >= MAX_ENTRIES) {
            entries.removeAt(0)
        }
        val entry = AuditEntry(eventType, method, requestId, detail)
        entries.add(entry)
        Logger.d(TAG, "Audit: $eventType - $method - $detail")
    }

    /**
     * Get all audit entries
     */
    fun getEntries(): List<AuditEntry> = entries.toList()

    /**
     * Get entries filtered by event type
     */
    fun getEntriesByType(eventType: String): List<AuditEntry> =
        entries.filter { it.eventType == eventType }

    /**
     * Get entries with limit
     */
    fun getEntries(limit: Int): List<AuditEntry> =
        entries.takeLast(limit)

    /**
     * Get summary: count per event type
     */
    fun getSummary(): Map<String, Int> {
        val summary = mutableMapOf<String, Int>()
        for (entry in entries) {
            summary[entry.eventType] = (summary[entry.eventType] ?: 0) + 1
        }
        return summary
    }

    /**
     * Clear all entries
     */
    fun clear() {
        entries.clear()
        Logger.d(TAG, "Audit log cleared")
    }
}

File: iOSWebBox/CoconutSDK/Sources/CoconutSDK/Bridge/SecurityAuditLog.swift
import Foundation

public class SecurityAuditLog: @unchecked Sendable {

    public static let shared = SecurityAuditLog()
    private let tag = "SecurityAuditLog"

    public static let EVENT_DOMAIN_REJECTED = "DOMAIN_REJECTED"
    public static let EVENT_RATE_LIMITED = "RATE_LIMITED"
    public static let EVENT_PARAMS_OVERSIZED = "PARAMS_OVERSIZED"
    public static let EVENT_SIGNATURE_INVALID = "SIGNATURE_INVALID"
    public static let EVENT_SIGNATURE_EXPIRED = "SIGNATURE_EXPIRED"
    public static let EVENT_NONCE_REUSED = "NONCE_REUSED"
    public static let EVENT_TOKEN_INVALID = "TOKEN_INVALID"

    public struct AuditEntry {
        public let eventType: String
        public let method: String
        public let requestId: String
        public let detail: String
        public let timestamp: Int64
    }

    private var entries: [AuditEntry] = []
    private let maxEntries = 500
    private let lock = NSLock()

    private init() {}

    public func record(eventType: String, method: String, requestId: String, detail: String) {
        lock.lock()
        defer { lock.unlock() }

        let entry = AuditEntry(
            eventType: eventType,
            method: method,
            requestId: requestId,
            detail: detail,
            timestamp: Int64(Date().timeIntervalSince1970 * 1000)
        )
        entries.append(entry)
        if entries.count > maxEntries {
            entries.removeFirst(entries.count - maxEntries)
        }
    }

    public func getEntries(limit: Int = 100) -> [AuditEntry] {
        lock.lock()
        defer { lock.unlock() }
        return Array(entries.suffix(limit))
    }

    public func getEntriesByType(_ eventType: String) -> [AuditEntry] {
        lock.lock()
        defer { lock.unlock() }
        return entries.filter { $0.eventType == eventType }
    }

    public func getSummary() -> [String: Any] {
        lock.lock()
        defer { lock.unlock() }

        var summary: [String: Int] = [:]
        for entry in entries {
            summary[entry.eventType, default: 0] += 1
        }
        let summaryList = summary.map { ["eventType": $0.key, "count": $0.value] as [String: Any] }
        return ["totalEvents": entries.count, "summary": summaryList]
    }

    public func clear() {
        lock.lock()
        defer { lock.unlock() }
        entries.removeAll()
    }
}
